packages/assets-controllers/src/DeFiPositionsController/DeFiPositionsController.ts

```diff
@@ -0,0 +1,159 @@
+import type { NetworkControllerStateChangeEvent } from '@metamask/network-controller';
+import type { Hex } from '@metamask/utils';
+
+import type { DefiPositionResponse } from './fetch-positions';
+import { buildPositionFetcher } from './fetch-positions';
+import { groupPositions, type GroupedPositions } from './group-positions';
+
+const controllerName = 'DeFiPositionsController';
+
+export type DeFiPositionsControllerState = {
+  /**
+   * Object containing DeFi positions per account and network
+   */
+  allDeFiPositions: {
+    [accountAddress: string]: { [chain: Hex]: GroupedPositions } | null;
+  };
+};
+
+const controllerMetadata: StateMetadata<DeFiPositionsControllerState> = {
+  allDeFiPositions: {
+    persist: false,
+    anonymous: false,
+  },
+};
+
+export const getDefaultDefiPositionsControllerState =
+  (): DeFiPositionsControllerState => {
+    return {
+      allDeFiPositions: {},
+    };
+  };
+
+export type DeFiPositionsControllerActions =
+  DeFiPositionsControllerGetStateAction;
+
+export type DeFiPositionsControllerGetStateAction = ControllerGetStateAction<
+  typeof controllerName,
+  DeFiPositionsControllerState
+>;
+
+export type DeFiPositionsControllerEvents =
+  DeFiPositionsControllerStateChangeEvent;
+
+export type DeFiPositionsControllerStateChangeEvent =
+  ControllerStateChangeEvent<
+    typeof controllerName,
+    DeFiPositionsControllerState
+  >;
+
+/**
+ * The external actions available to the {@link DeFiPositionsController}.
+ */
+export type AllowedActions = AccountsControllerGetSelectedAccountAction;
+
+/**
+ * The external events available to the {@link DeFiPositionsController}.
+ */
+export type AllowedEvents =
+  | NetworkControllerStateChangeEvent
+  | AccountsControllerSelectedAccountChangeEvent;
+
+/**
+ * The messenger of the {@link DeFiPositionsController}.
+ */
+export type DeFiPositionsControllerMessenger = RestrictedMessenger<
+  typeof controllerName,
+  DeFiPositionsControllerActions | AllowedActions,
+  DeFiPositionsControllerEvents | AllowedEvents,
+  AllowedActions['type'],
+  AllowedEvents['type']
+>;
+
+/**
+ * Controller that stores assets and exposes convenience methods
+ */
+export class DeFiPositionsController extends BaseController<
+  typeof controllerName,
+  DeFiPositionsControllerState,
+  DeFiPositionsControllerMessenger
+> {
+  readonly #fetchPositions: (
+    accountAddress: string,
+  ) => Promise<DefiPositionResponse[]>;
+
+  /**
+   * Tokens controller options
+   *
+   * @param options - Constructor options.
+   * @param options.messenger - The controller messenger.
+   * @param options.state - Initial state to set on this controller.
+   * @param options.apiUrl - Override for theAPI URL to use for fetching DeFi positions.
+   */
+  constructor({
+    messenger,
+    state,
+    apiUrl,
+  }: {
+    messenger: DeFiPositionsControllerMessenger;
+    state?: Partial<DeFiPositionsControllerState>;
+    apiUrl?: string;
```

Comment:
Do we need to be able to turn this controller/fetch calls off when "Basic Functionality" is disabled?

We can do this by injecting a lazy `isEnabled()` property into the controller
```ts
// inside extension
const controller = new DeFiPositionsController({
  ...,
  isEnabled: () => this.preferencesController.state.useExternalServices === true,
})
```

I'm not sure if this controller is shared or different on Mobile. If they are shared, then maybe we can use messaging system?

Reply:
This is probably relevant and I wasn't aware of it, thanks. I'll add the switch to disable the api calls

Reply:
I have added an `isEnabled: () => boolean` input to the constructor. Default to `() => true`.

This allows the client to specify the conditions, which can be a mixture of state from another controller and a feature flag.

Reply:
So it turns we cannot do that, and it's probably a good thing. With the new initialisation method we no longer have access to `this.preferencesController.state` from the init function.

It is probably best to use the intended way, which is for `DeFiPositionsController` to query `PreferencesController` state through a call.

Reply:
Amazing work.

Yeah hopefully extension and mobile centralise their preference controller that way we can use messaging system events/actions.